### PR TITLE
feat(execution-engine)!: stream map to scalar conversion using canon instruction [fixes VM-294]

### PR DESCRIPTION
**Pull request by raftedproc**
feat(execution-engine): Stream Map to Scalar conversion using canon instruction [fixes VM-294]

### CODE REVIEW

File: air/src/execution_step/boxed_value/stream.rs [hunk 1]
```diff
@@ -211,6 +215,28 @@ impl Stream {
     fn remove_empty_generations(&mut self) {
         self.values.retain(|values| !values.is_empty());
     }
+
+    pub(crate) fn compactify_stream_map_scalar(&self) -> Cow<'_, Stream> {
```
[the remark this answers is not in the document]
Comment (mikevoronov) [reply air/src/execution_step/boxed_value/stream.rs:219]: And it's better to add unit tests for this function.

File: air/src/execution_step/instructions/canon.rs [hunk 2]
```diff
@@ -40,17 +42,56 @@ impl<'i> super::ExecutableInstruction<'i> for ast::Canon<'i> {
         log_instruction!(call, exec_ctx, trace_ctx);
         let canon_result = trace_to_exec_err!(trace_ctx.meet_canon_start(), self)?;
 
+        let epilog: &dyn for<'ctx> Fn(
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon.rs:45]: Let's introduce a separate type for it.

File: air/src/execution_step/instructions/canon.rs [hunk 3]
```diff
@@ -74,15 +115,21 @@ fn handle_seen_canon(
...
 
-fn handle_unseen_canon(
-    ast_canon: &ast::Canon<'_>,
+pub(super) fn handle_unseen_canon(
+    epilog: &dyn for<'ctx> Fn(
+        StreamWithSerializedView,
+        &mut ExecutionCtx<'_>,
+        &mut TraceHandler,
+    ) -> ExecutionResult<()>,
+    get_stream_or_default: impl for<'ctx> Fn(&'ctx mut ExecutionCtx<'_>) -> Cow<'ctx, Stream>,
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon.rs:127]: Let's name it just `get_stream_or_default`.

File: air/src/execution_step/instructions/canon.rs [hunk 4]
```diff
@@ -40,17 +42,56 @@ impl<'i> super::ExecutableInstruction<'i> for ast::Canon<'i> {
...
         match canon_result {
             MergerCanonResult::CanonResult(canon_result_cid) => {
-                handle_seen_canon(self, canon_result_cid, exec_ctx, trace_ctx)
+                handle_seen_canon(epilog, canon_result_cid, exec_ctx, trace_ctx)
+            }
+            MergerCanonResult::Empty => {
+                // This closure gets a stream from a context or returns a default empty stream,
+                // it is crucial for deterministic behaviour, for more info see
+                // github.com/fluencelabs/aquavm/issues/346
+                let get_stream_or_default: &dyn for<'ctx> Fn(&'ctx mut ExecutionCtx<'_>) -> Cow<'ctx, Stream> =
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon.rs:75]: For better code readability, I suggest to leaving this as a separate function that will return a closure.

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            let value = JValuable::as_jvalue(&&canon_stream).into_owned();
+            let tetraplet = canon_stream.tetraplet().clone();
+            let position = trace_ctx.trace_pos().map_err(UncatchableError::from)?;
+            let value = CanonResultAggregate::new(
+                Rc::new(value),
+                tetraplet.peer_pk.as_str().into(),
+                &tetraplet.json_path,
+                position,
+            );
+            let result: ValueAggregate = ValueAggregate::from_canon_result(value, canon_result_cid.clone());
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon_stream_map_scalar.rs:66]:
```suggestion
            let result = ValueAggregate::from_canon_result(value, canon_result_cid.clone());
```

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+impl<'i> super::ExecutableInstruction<'i> for ast::CanonStreamMapScalar<'i> {
+    #[tracing::instrument(level = "debug", skip(exec_ctx, trace_ctx))]
+    fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
+        log_instruction!(call, exec_ctx, trace_ctx);
+        let canon_result = trace_to_exec_err!(trace_ctx.meet_canon_start(), self)?;
+
+        let epilog: &dyn for<'ctx> Fn(
+            StreamWithSerializedView,
+            &mut ExecutionCtx<'_>,
+            &mut TraceHandler,
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon_stream_map_scalar.rs:47]: ditto with a separate type

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        match canon_result {
+            MergerCanonResult::CanonResult(canon_result_cid) => {
+                handle_seen_canon(epilog, canon_result_cid, exec_ctx, trace_ctx)
+            }
+            MergerCanonResult::Empty => {
+                // This closure gets underlying stream from a StreamMap in a context
+                // or returns a default empty stream,
+                // it is crucial for deterministic behaviour, for more info see
+                // github.com/fluencelabs/aquavm/issues/346
+                let get_stream_or_default: &dyn for<'ctx> Fn(&'ctx mut ExecutionCtx<'_>) -> Cow<'ctx, Stream> =
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon_stream_map_scalar.rs:82]: ditto with a separate function

File: air/src/execution_step/boxed_value/stream.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (raftedproc) [reply air/src/execution_step/boxed_value/stream.rs:219]: Map call that is at the line 228 uses distinct_values to collect unique keys.

File: air/src/execution_step/instructions/canon.rs (hunk 4 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/canon.rs:75]: Pardon my ignorance, but I am somewhat skeptical that a closure that returns Fn trait object or a Box improves readability :)
Will try it though

File: air/src/execution_step/boxed_value/stream.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (raftedproc) [reply air/src/execution_step/boxed_value/stream.rs]: Done.

File: air/src/execution_step/instructions/canon.rs (hunk 2 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/canon.rs]: Done

File: air/src/execution_step/instructions/canon.rs (hunk 4 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/canon.rs]: Done

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs (hunk 6 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/canon_stream_map_scalar.rs]: Done

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs (hunk 7 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/canon_stream_map_scalar.rs]: Done

File: air/src/execution_step/instructions/canon.rs [hunk 8]
```diff
@@ -29,28 +30,55 @@ use air_interpreter_data::CanonCidAggregate;
...
+                canon_result_cid,
+            } = stream_with_positions;
+
+            exec_ctx.scalars.set_canon_value(
+                self.canon_stream.name,
+                CanonStreamWithProvenance::new(canon_stream, canon_result_cid.clone()),
+            )?;
+
+            trace_ctx.meet_canon_end(CanonResult::new(canon_result_cid));
+            Ok(())
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon.rs:64]:
```suggestion
            let value = CanonStreamWithProvenance::new(canon_stream, canon_result_cid.clone());
            exec_ctx.scalars.set_canon_value(self.canon_stream.name, value)?;
            trace_ctx.meet_canon_end(CanonResult::new(canon_result_cid));

            Ok(())
```

File: air/src/execution_step/instructions/canon_stream_map_scalar.rs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                handle_unseen_canon(epilog, &get_stream_or_default, &self.peer_id, exec_ctx, trace_ctx)
+            }
+        }
+    }
+}
+
+/// The resulting closure gets underlying stream from a StreamMap in a context
+/// or returns a default empty stream,
+/// it is crucial for deterministic behaviour, for more info see
+/// github.com/fluencelabs/aquavm/issues/346
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon_stream_map_scalar.rs:88]: Let's add a dot to the end of this comment according to our [standard](https://github.com/fluencelabs/aquavm/blob/master/docs/developer-notes.md#coding-style).

File: air/src/execution_step/instructions/canon.rs [hunk 10]
```diff
@@ -175,10 +185,21 @@ fn create_canon_stream_from_name(
...
 }
 
-/// This function gets a stream from context or return a default empty stream,
-/// it's crucial for deterministic behaviour, for more info see
+pub(super) type GetStreamClosure<'obj> = dyn for<'ctx> Fn(&'ctx mut ExecutionCtx<'_>) -> Cow<'ctx, Stream> + 'obj;
+
+/// The resulting closure gets underlying stream in a context
+/// or returns a default empty stream,
+/// it is crucial for deterministic behaviour, for more info see
 /// github.com/fluencelabs/aquavm/issues/346
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/canon.rs:193]: ditto with dot

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**